physical/spanner/spanner.go

```diff
@@ -0,0 +1,342 @@
+	// database name.
+	envDatabase = "GOOGLE_SPANNER_DATABASE"
+
+	// envHAEnabled is the name of the environment variable to search for the
+	// boolean indicating if HA is enabled.
+	envHAEnabled = "GOOGLE_SPANNER_HA_ENABLED"
+
+	// envHATable is the name of the environment variable to search for the table
+	// name to use for HA.
+	envHATable = "GOOGLE_SPANNER_HA_TABLE"
+
+	// envTable is the name of the environment variable to search for the table
+	// name.
+	envTable = "GOOGLE_SPANNER_TABLE"
+
+	// defaultTable is the default table name if none is specified.
+	defaultTable = "Vault"
+
+	// defaultHASuffix is the default suffix to apply to the table name if no
+	// HA table is provided.
+	defaultHASuffix = "HA"
+)
+
+var (
+	// metricDelete is the key for the metric for measuring a Delete call.
+	metricDelete = []string{"spanner", "delete"}
+
+	// metricGet is the key for the metric for measuring a Get call.
+	metricGet = []string{"spanner", "get"}
+
+	// metricList is the key for the metric for measuring a List call.
+	metricList = []string{"spanner", "list"}
+
+	// metricPut is the key for the metric for measuring a Put call.
+	metricPut = []string{"spanner", "put"}
+)
+
+// Backend implements physical.Backend and describes the steps necessary to
+// persist data using Google Cloud Spanner.
+type Backend struct {
+	// database is the name of the database to use for data storage and retrieval.
+	// This is supplied as part of user configuration.
+	database string
+
+	// table is the name of the table in the database.
+	table string
+
+	// haTable is the name of the table to use for HA in the database.
+	haTable string
+
+	// haEnabled indicates if high availability is enabled. Default: true.
+	haEnabled bool
+
+	// client is the underlying API client for talking to spanner.
+	client *spanner.Client
+
+	// logger and permitPool are internal constructs.
+	logger     log.Logger
+	permitPool *physical.PermitPool
+}
+
+// NewBackend creates a new Google Spanner storage backend with the given
+// configuration. This uses the official Golang Cloud SDK and therefore supports
+// specifying credentials via envvars, credential files, etc.
+func NewBackend(c map[string]string, logger log.Logger) (physical.Backend, error) {
+	logger.Debug("physical/spanner: configuring backend")
+
+	// Database name
+	database := os.Getenv(envDatabase)
+	if database == "" {
+		database = c["database"]
+	}
+	if database == "" {
+		return nil, errors.New("missing database name")
+	}
+
+	// Table name
+	table := os.Getenv(envTable)
+	if table == "" {
+		table = c["table"]
+	}
+	if table == "" {
+		table = defaultTable
+	}
+
+	// HA table name
+	haTable := os.Getenv(envHATable)
+	if haTable == "" {
+		haTable = c["ha_table"]
+	}
+	if haTable == "" {
+		haTable = table + defaultHASuffix
+	}
+
+	// HA configuration
+	haEnabledStr := os.Getenv(envHAEnabled)
+	if haEnabledStr == "" {
+		haEnabledStr = c["ha_enabled"]
+	}
+	haEnabled := true
```

Comment:
Please default to false. We don't test this and we've had a lot of people with other HA backends that end up having some kind of problem. This isn't to say that this backend will, only that we would prefer people need to opt-in, at least in the near term.

Reply:
Done